### PR TITLE
Implement Sharpe ratio Mod

### PR DESCRIPTION
**Pull request by etowle**
### Description

Implements Mod #15.

### Checklist

- Implementation:
  - [x] Implementation of the Mod in the `gurobi_optimods` installable package
  - [x] Tests for the Mod implementation in `tests/`
  - [x] Docstrings for public API, correctly linked using sphinx-autodoc
- Documentation page:
  - [x] Background and problem specification
  - [x] Example of the input data format (use `gurobi_optimods.datasets` for loading data)
  - [x] Runnable code example
  - [x] Presentation of solutions
  - [x] Included in the mod gallery and toctree

### CODE REVIEW

File: docs/source/mods/sharpe-ratio.rst [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+
+The Sharpe ratio, named after Nobel laureate William F. Sharpe, measures the risk-adjusted performance of a portfolio. It is calculated by subtracting the risk-free rate of return from a portfolio's expected return and dividing the result by the portfolio's standard deviation. Thus, a higher Sharpe ratio corresponds to more attractive risk-adjusted performance, as the portfolio's returns are higher relative to the portfolio's risk.
+
+This OptiMod computes the portfolio that maximizes the Sharpe ratio for the given expected returns and variances.
+
+
+Problem Specification
+---------------------
+
+.. tabs::
```
Comment (simonbowly) [inline docs/source/mods/sharpe-ratio.rst:12]: Can you tweak this section to use the "dropdown" directive for the mathsy parts (instead of tabs)? Check the bipartite matching mod for an example.

File: docs/source/mods/sharpe-ratio.rst [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+            \end{alignat*}
+
+        Model :math:`\textbf{(3)}` is convex. In this OptiMod, we solve the convex quadratic program (QP) :math:`\textbf{(3)}`, then map the optimal solution :math:`y^*` back to the original problem :math:`\textbf{(1)}` via the transformation :math:`x^*_i := y^*_i / \sum_{j=1}^n y^*_j` for :math:`i = 1, \ldots, n`.
+
+Interface
+---------
+
+The ``max_sharpe_ratio`` function requires the following arguments:
+
+* ``cov_matrix``: The covariance matrix :math:`\Sigma`, given as a numpy ndarray or pandas DataFrame (e.g., returned by the `pandas.DataFrame.cov <https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.cov.html>`_ method).
```
Comment (simonbowly) [inline docs/source/mods/sharpe-ratio.rst:64]: You can get this link automagically from intersphinx. Either::

    :meth:`DataFrame.cov <pandas.DataFrame.cov>`

or just::

    :meth:`pandas.DataFrame.cov`

depending how you want it to appear.

File: src/gurobi_optimods/sharpe_ratio.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        raise ValueError(
+            f"No expected returns are greater than risk-free return rate of {rf_rate}"
+        )
+
+    portfolio, ratio = _max_sharpe_ratio_numpy(cov_matrix, mu, rf_rate, create_env)
+
+    if indices is not None:
+        portfolio = pd.Series(data=portfolio, index=indices)
+
+    return portfolio, ratio
```
Comment (simonbowly) [inline src/gurobi_optimods/sharpe_ratio.py:77]: Could return a dataclass containing the two elements, instead of a tuple? I am not sure which is better...

**Review by simonbowly — COMMENTED** [review]
Looking great @etowle! I updated the docstrings to numpydoc style and fixed the CI formatting complaints. Other comments below.

**Comment by simonbowly** [comment]
@rluce can you take a look through the docs when you have some time? It's built at: https://gurobi-optimization-gurobi-optimods--97.com.readthedocs.build/en/97/mods/sharpe-ratio.html

File: src/gurobi_optimods/sharpe_ratio.py (hunk 3 above)
Comment (etowle) [reply src/gurobi_optimods/sharpe_ratio.py]: I was on the fence about this. I chose the tuple because I was only returning two objects. However, looking at the Mean-Variance Portfolio Mod, I like the idea of also returning the estimated return and risk, which is better done via a data class.

I created a new `SharpeRatioResult` data class and changed the return type.

**Comment by rluce** [comment]
@silkehorn Could you please review this Mod?

File: docs/source/mods/sharpe-ratio.rst [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+    .. math::
+        \begin{alignat*}{2}
+            \max_y\ && \frac{1}{\sqrt{y^\top \Sigma y}} \qquad & \\
+            \textrm{s.t.}\enspace && (\mu - r_f)^\top y = {} &1 \qquad \textbf{(2)} \\
+            && y \geq {} &0.
+        \end{alignat*}
+
+    Models :math:`\textbf{(1)}` and :math:`\textbf{(2)}` are equivalent in the sense that given a solution to either problem, we can construct a solution to the other of equal or better objective value. In particular, any solution :math:`\bar{x}` of :math:`\textbf{(1)}` can be mapped to a solution :math:`\bar{y}` of :math:`\textbf{(2)}` of equivalent objective value using the transformation :math:`\bar{y}_i := \bar{x}_i / \mu^\top \bar{x}` for :math:`i = 1, \ldots, n`. Conversely, any solution :math:`\bar{y}` of :math:`\textbf{(2)}` can be mapped to a solution :math:`\bar{x}` of :math:`\textbf{(1)}` of equivalent objective value using the transformation :math:`\bar{x}_i := \bar{y}_i / \sum_{j = 1}^n \bar{y}_j` for :math:`i = 1, \ldots, n`.
```
Comment (silkehorn) [inline docs/source/mods/sharpe-ratio.rst:41]: Even though there is no mathematical ambiguity, I find fractions with `/` sometimes hard to parse. 
For example, I would find $\frac{\bar{x}_i}{\mu^\top \bar{x}}$ easier to read than $\bar{x}_i / \mu^\top \bar{x}$. This has the disadvantage that the font gets very small, though. So both versions have their merits, and I don't have a strong opinion on this.

File: docs/source/mods/sharpe-ratio.rst [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+    Like :math:`\textbf{(1)}`, model :math:`\textbf{(2)}` is non-convex. However, because :math:`\Sigma` is positive semidefinite, maximizing :math:`1 / \sqrt{y^\top \Sigma y}` is equivalent to minimizing :math:`y^\top \Sigma y`. Thus, the optimal solution of :math:`\textbf{(2)}` is equivalent to the optimal solution of the following model:
+
+    .. math::
+        \begin{alignat*}{2}
+            \min_y\ && y^\top \Sigma y \qquad \quad & \\
+            \textrm{s.t.}\enspace && (\mu - r_f)^\top y = {} &1 \qquad \textbf{(3)} \\
+            && y \geq {} &0.
+        \end{alignat*}
+
+    Model :math:`\textbf{(3)}` is convex. In this OptiMod, we solve the convex quadratic program (QP) :math:`\textbf{(3)}`, then map the optimal solution :math:`y^*` back to the original problem :math:`\textbf{(1)}` via the transformation :math:`x^*_i := y^*_i / \sum_{j=1}^n y^*_j` for :math:`i = 1, \ldots, n`.
```
Comment (silkehorn) [inline docs/source/mods/sharpe-ratio.rst:52]: Same as above, but here I think that the final transformation is important enough to warrant display rather than inline style. Then it could also be a "real" fraction without the small font problem.

**Review by silkehorn — COMMENTED** [review]
This is great! 
* The implementation is very clear and concise.
* The documentation (in particular, the mathematical background) is well written. The comparison with the Monte Carlo method is also very interesting. 
* The gallery image is really cool!
* I also like how the argument names align with the portfolio mod. 

I only made two small comments to the documentation.

File: docs/source/mods/sharpe-ratio.rst (hunk 4 above)
Comment (etowle) [reply docs/source/mods/sharpe-ratio.rst:41]: I have a slight preference for using `/` instead of `\frac` for inline fractions here. When using `\frac`, the text is too small for me to comfortably read.

inline_frac

File: docs/source/mods/sharpe-ratio.rst (hunk 5 above)
Comment (etowle) [reply docs/source/mods/sharpe-ratio.rst]: Good idea. I moved the final transformation to display style.

**Comment by etowle** [comment]
Thanks for the reviews, @silkehorn and @simonbowly! I've made changes based on your comments.

**Comment by simonbowly** [comment]
Thanks for reviewing @silkehorn and thanks for the great work on this @etowle! Looks like we're all set, I'll merge it shortly 👍
